File: windowsApp/src/Rally.Core/DiskCache.cs
using System.Text.Json;

// Atomic JSON disk cache with TTL. Mirrors DiskCache.swift: envelope with
// timestamp + payload, temp-dir injectable for tests.
namespace Rally.Core;

public sealed class DiskCache(string? directory = null)
{
    private readonly string _dir = directory
        ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Rally", "cache");

    private string PathFor(string name) => Path.Combine(_dir, name);

    public T? Load<T>(string name, TimeSpan maxAge)
    {
        try
        {
            Directory.CreateDirectory(_dir);
            var path = PathFor(name);
            if (!File.Exists(path)) return default;
            using var doc = JsonDocument.Parse(File.ReadAllText(path));
            var root = doc.RootElement;
            if (!root.TryGetProperty("savedAt", out var saved) || !saved.TryGetInt64(out var ms))
                return default;
            if (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - ms > maxAge.TotalMilliseconds)
                return default;
            if (!root.TryGetProperty("payload", out var payload)) return default;
            return payload.Deserialize<T>();
        }
        catch { return default; }
    }

    public T? LoadAny<T>(string name)
    {
        try
        {
            var path = PathFor(name);
            if (!File.Exists(path)) return default;
            using var doc = JsonDocument.Parse(File.ReadAllText(path));
            if (!doc.RootElement.TryGetProperty("payload", out var payload)) return default;
            return payload.Deserialize<T>();
        }
        catch { return default; }
    }

    public void Save<T>(T value, string name)
    {
        try
        {
            Directory.CreateDirectory(_dir);
            var tmp = PathFor(name + ".tmp");
            var json = JsonSerializer.Serialize(new { savedAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(), payload = value });
            File.WriteAllText(tmp, json);
            File.Move(tmp, PathFor(name), overwrite: true);
        }
        catch { }
    }
}

// Last-known-good ESPN schedule. 72h TTL; stale data beats an empty shelf.
public sealed class ScheduleStore(string? directory = null)
{
    public static readonly TimeSpan Ttl = TimeSpan.FromHours(72);
    private const string Name = "schedule.json";
    private readonly DiskCache _cache = new(directory);

    public List<SportEvent>? LoadFresh()
    {
        var events = _cache.Load<List<SportEvent>>(Name, Ttl);
        return events is { Count: > 0 } ? events : null;
    }

    public List<SportEvent>? LoadAny()
    {
        var events = _cache.LoadAny<List<SportEvent>>(Name);
        return events is { Count: > 0 } ? events : null;
    }

    public void Save(List<SportEvent> events)
    {
        if (events.Count == 0) return;
        _cache.Save(events, Name);
    }
}

// Persisted IPTV catalog, identity-gated like the in-memory cache.
public sealed class ChannelDiskStore(string? directory = null)
{
    public static readonly TimeSpan Ttl = TimeSpan.FromMinutes(15);
    private readonly DiskCache _cache = new(directory);

    public static string FileName(string identity)
    {
        var safe = new string(identity.ToLowerInvariant()
            .Select(c => char.IsLetterOrDigit(c) ? c : '_').ToArray());
        if (safe.Length > 64) safe = safe[..64];
        return $"channels-{safe}.json";
    }

    public List<IptvChannel>? LoadFresh(string identity)
    {
        var channels = _cache.Load<List<IptvChannel>>(FileName(identity), Ttl);
        return channels is { Count: > 0 } ? channels : null;
    }

    public List<IptvChannel>? LoadAny(string identity)
    {
        var channels = _cache.LoadAny<List<IptvChannel>>(FileName(identity));
        return channels is { Count: > 0 } ? channels : null;
    }

    public void Save(List<IptvChannel> channels, string identity)
    {
        if (channels.Count == 0) return;
        _cache.Save(channels, FileName(identity));
    }
}

File: appleApp/Sources/RallyCore/DiskCache.swift
import Foundation

/// Atomic JSON disk cache with TTL. boring foundation for the schedule +
/// channel stores below (mirrors SportsEventDiskCache / Room serve-instantly).
public struct DiskCache: Sendable {
    public let directory: URL

    public init(directory: URL? = nil) {
        if let directory {
            self.directory = directory
        } else {
            let base = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!
            self.directory = base.appendingPathComponent("com.shiv.rally.macos", isDirectory: true)
        }
        try? FileManager.default.createDirectory(at: self.directory, withIntermediateDirectories: true)
    }

    private struct Envelope<T: Codable>: Codable {
        var savedAt: Date
        var payload: T
    }

    public func load<T: Codable>(_ name: String, maxAge: TimeInterval) -> T? {
        guard let data = try? Data(contentsOf: directory.appendingPathComponent(name)),
              let env = try? JSONDecoder().decode(Envelope<T>.self, from: data),
              Date().timeIntervalSince(env.savedAt) < maxAge else { return nil }
        return env.payload
    }

    public func loadAny<T: Codable>(_ name: String) -> T? {
        guard let data = try? Data(contentsOf: directory.appendingPathComponent(name)),
              let env = try? JSONDecoder().decode(Envelope<T>.self, from: data) else { return nil }
        return env.payload
    }

    public func save<T: Codable>(_ value: T, name: String) {
        let env = Envelope(savedAt: Date(), payload: value)
        guard let data = try? JSONEncoder().encode(env) else { return }
        // Atomic write: no torn reads after a crash or relaunch mid-save.
        try? data.write(to: directory.appendingPathComponent(name), options: .atomic)
    }
}

/// Last-known-good ESPN schedule. 72h TTL; stale data still beats an empty
/// shelf on airplane-mode/DNS-outage launches (Android SportsEventDiskCache).
public struct ScheduleStore: Sendable {
    public static let ttl: TimeInterval = 72 * 3600
    private static let name = "schedule.json"
    private let cache: DiskCache

    public init(directory: URL? = nil) {
        cache = DiskCache(directory: directory)
    }

    public func loadFresh() -> [SportEvent]? {
        let events: [SportEvent]? = cache.load(Self.name, maxAge: Self.ttl)
        return (events?.isEmpty == false) ? events : nil
    }

    public func loadAny() -> [SportEvent]? {
        let events: [SportEvent]? = cache.loadAny(Self.name)
        return (events?.isEmpty == false) ? events : nil
    }

    public func save(_ events: [SportEvent]) {
        guard !events.isEmpty else { return }
        cache.save(events, name: Self.name)
    }
}

/// Persisted IPTV channel catalog, identity-gated like the in-memory cache
/// (portal|mac or xtream identity). 15-min fresh TTL; stale fallback keeps
/// Live TV usable when the portal is briefly unreachable.
public struct ChannelDiskStore: Sendable {
    public static let ttl: TimeInterval = 15 * 60
    private let cache: DiskCache

    public init(directory: URL? = nil) {
        cache = DiskCache(directory: directory)
    }

    public static func fileName(identity: String) -> String {
        let safe = identity.lowercased().map { $0.isLetter || $0.isNumber ? String($0) : "_" }.joined()
        return "channels-\(safe.prefix(64)).json"
    }

    public func loadFresh(identity: String) -> [IptvChannel]? {
        let channels: [IptvChannel]? = cache.load(Self.fileName(identity: identity), maxAge: Self.ttl)
        return (channels?.isEmpty == false) ? channels : nil
    }

    public func loadAny(identity: String) -> [IptvChannel]? {
        let channels: [IptvChannel]? = cache.loadAny(Self.fileName(identity: identity))
        return (channels?.isEmpty == false) ? channels : nil
    }

    public func save(_ channels: [IptvChannel], identity: String) {
        guard !channels.isEmpty else { return }
        cache.save(channels, name: Self.fileName(identity: identity))
    }
}
